cmd/kubeadm/app/discovery/file/file.go

```diff
@@ -17,13 +17,96 @@ limitations under the License.
 package file
 
 import (
+	"fmt"
 	"net/url"
 
-	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+	bootstrapapi "k8s.io/kubernetes/pkg/bootstrap/api"
 )
 
-func Parse(u *url.URL, c *kubeadm.Discovery) {
-	c.File = &kubeadm.FileDiscovery{
+func Parse(u *url.URL, c *kubeadmapi.Discovery) {
+	c.File = &kubeadmapi.FileDiscovery{
 		Path: u.Path,
 	}
 }
+
+// RetrieveValidatedClusterInfo connects to the API Server and makes sure it can talk
+// securely to the API Server using the provided CA cert and
+// optionally refreshes the cluster-info information from the cluster-info ConfigMap
+func RetrieveValidatedClusterInfo(d *kubeadmapi.FileDiscovery) (*clientcmdapi.Cluster, error) {
+	clusterinfo, err := clientcmd.LoadFromFile(d.Path)
+	if err != nil {
+		return nil, err
+	}
+	return ValidateClusterInfo(clusterinfo)
+}
+
+// ValidateClusterInfo connects to the API Server and makes sure it can talk
+// securely to the API Server using the provided CA cert and
+// optionally refreshes the cluster-info information from the cluster-info ConfigMap
+func ValidateClusterInfo(clusterinfo *clientcmdapi.Config) (*clientcmdapi.Cluster, error) {
+	client, err := kubeconfigutil.KubeConfigToClientSet(clusterinfo)
+	if err != nil {
+		return nil, err
+	}
+
+	// This is the cluster object we've got from
+	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(clusterinfo)
+
+	fmt.Printf("[discovery] Created cluster-info discovery client, requesting info from %q\n", defaultCluster.Server)
+
+	var clusterinfoCM *v1.ConfigMap
+	wait.PollInfinite(constants.DiscoveryRetryInterval, func() (bool, error) {
```

Comment:
Are we sure we want PollInfinite here?  If the cert is bad, it'll never exit.  What is the thinking there?